### PR TITLE
Analyze PSBT function

### LINKED ISSUE #2222: Allow ability to `analyzepsbt`
Given a PSBT it seems I have no way to look at the contents of it?

https://developer.bitcoin.org/reference/rpc/analyzepsbt.html

https://bitcoin-s.org/docs/next/applications/server#psbt

### PR DESCRIPTION
**Pull request by benthecarman**
Fixes #2222

### CODE REVIEW

**Review by Christewart — CHANGES_REQUESTED** [review]
tests needed as well i believe?

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala [hunk 1]
```diff
@@ -154,6 +154,113 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
...
+      case locktime: LockTimeScriptPubKey =>
+        missingSigsFromScript(locktime.nestedScriptPubKey)
+      case cond: ConditionalScriptPubKey =>
+        val first = missingSigsFromScript(cond.firstSPK)
+        val second = missingSigsFromScript(cond.secondSPK)
+        first ++ second
+    }
+  }
+
+  def missingSignatures(vout: Int): Vector[Sha256Hash160Digest] = {
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:264]: scaladoc needed, I have no clue what this is doing and why it hashes pubkeys that don't seem to need to be hashed

**Comment by Christewart** [comment]
Ok this seems to work, first here is the PSBT

>cHNidP8BAHECAAAAAaPTWHbUU9qZH3E8kIkbibcteaNNTH6csUa+pkkyrqdQAAAAAAD+////AgCHkwMAAAAAFgAU0vJ6xEpYLZgq0XT4Lxnrz3UcBhsAWmICAAAAABYAFMwT65eVs7MNdEmZ8a/btlp3iIuzAAAAAAABAP1AAQIAAAAFKZIoajB5IH5z9iHokxOlX2rQhwgcCyzaDPXXWbR1tJoAAAAAAP////9pTi0Pa6WrpNiaA3pqZ+YBse+v1jgpHtTzZy/vQHF0gkYAAAAA/////xfvpgcSWII60dKIxuOFbZoOZXgEMzh46FPKi3ciQ76kNwAAAAD//////mFLtbzVbA5rw0pzjzJPZP2fFf8+hHcPx/DsUmw9zSdRAAAAAP/////aFYTyhqQ1D5+3KyYhJklDPVsM7TZ5ckF+Ki2GCREY6TwAAAAA/////wOa5/UFAAAAACIAIPrv3JDIVLeft5JpaP948L/cyobLz9N9fzvM5qNeV34tLGwhAQAAAAAWABQyQn7LBTqGwlw1YR4czWqI99wxNk9PLAAAAAAAFgAUmp9tVa99KP5hNYiVoqw9Q20ISukAAAAAAQErmuf1BQAAAAAiACD679yQyFS3n7eSaWj/ePC/3MqGy8/TfX87zOajXld+LSICAnuk3VLqOOGKWaQyBUAiA0Dkh9fJlvN+voAZ2YE4q6lGRzBEAiBgJOvBFUyD1zFe1cCUi7Q+mabqnDPGwgJMVUzTM+9sqgIgYHaAZKgPoEzzDEc3RZYG1Sxsb4n3eM9OEwpaQaWieTABAQVHUiEDgsVFqH1kh8VRENc3ORR0Nyf7G4DuBuTJxi4ln6dRMTMhAnuk3VLqOOGKWaQyBUAiA0Dkh9fJlvN+voAZ2YE4q6lGUq4AAAA=


### bitcoin core results

>  ./bitcoin-cli analyzepsbt cHNidP8BAHECAAAAAaPTWHbUU9qZH3E8kIkbibcteaNNTH6csUa+pkkyrqdQAAAAAAD+////AgCHkwMAAAAAFgAU0vJ6xEpYLZgq0XT4Lxnrz3UcBhsAWmICAAAAABYAFMwT65eVs7MNdEmZ8a/btlp3iIuzAAAAAAABAP1AAQIAAAAFKZIoajB5IH5z9iHokxOlX2rQhwgcCyzaDPXXWbR1tJoAAAAAAP////9pTi0Pa6WrpNiaA3pqZ+YBse+v1jgpHtTzZy/vQHF0gkYAAAAA/////xfvpgcSWII60dKIxuOFbZoOZXgEMzh46FPKi3ciQ76kNwAAAAD//////mFLtbzVbA5rw0pzjzJPZP2fFf8+hHcPx/DsUmw9zSdRAAAAAP/////aFYTyhqQ1D5+3KyYhJklDPVsM7TZ5ckF+Ki2GCREY6TwAAAAA/////wOa5/UFAAAAACIAIPrv3JDIVLeft5JpaP948L/cyobLz9N9fzvM5qNeV34tLGwhAQAAAAAWABQyQn7LBTqGwlw1YR4czWqI99wxNk9PLAAAAAAAFgAUmp9tVa99KP5hNYiVoqw9Q20ISukAAAAAAQErmuf1BQAAAAAiACD679yQyFS3n7eSaWj/ePC/3MqGy8/TfX87zOajXld+LSICAnuk3VLqOOGKWaQyBUAiA0Dkh9fJlvN+voAZ2YE4q6lGRzBEAiBgJOvBFUyD1zFe1cCUi7Q+mabqnDPGwgJMVUzTM+9sqgIgYHaAZKgPoEzzDEc3RZYG1Sxsb4n3eM9OEwpaQaWieTABAQVHUiEDgsVFqH1kh8VRENc3ORR0Nyf7G4DuBuTJxi4ln6dRMTMhAnuk3VLqOOGKWaQyBUAiA0Dkh9fJlvN+voAZ2YE4q6lGUq4AAAA=

>{
  "inputs": [
    {
      "has_utxo": true,
      "is_final": false,
      "next": "signer",
      "missing": {
        "signatures": [
          "4ad7e2fc048e440a9dc0b8bc7751df6b04e11cf0"
        ]
      }
    }
  ],
  "estimated_vsize": 168,
  "estimated_feerate": 0.00010059,
  "fee": 0.00001690,
  "next": "signer"
}


### bitcoin-s results 

>./bitcoin-s-cli analyzepsbt cHNidP8BAHECAAAAAaPTWHbUU9qZH3E8kIkbibcteaNNTH6csUa+pkkyrqdQAAAAAAD+////AgCHkwMAAAAAFgAU0vJ6xEpYLZgq0XT4Lxnrz3UcBhsAWmICAAAAABYAFMwT65eVs7MNdEmZ8a/btlp3iIuzAAAAAAABAP1AAQIAAAAFKZIoajB5IH5z9iHokxOlX2rQhwgcCyzaDPXXWbR1tJoAAAAAAP////9pTi0Pa6WrpNiaA3pqZ+YBse+v1jgpHtTzZy/vQHF0gkYAAAAA/////xfvpgcSWII60dKIxuOFbZoOZXgEMzh46FPKi3ciQ76kNwAAAAD//////mFLtbzVbA5rw0pzjzJPZP2fFf8+hHcPx/DsUmw9zSdRAAAAAP/////aFYTyhqQ1D5+3KyYhJklDPVsM7TZ5ckF+Ki2GCREY6TwAAAAA/////wOa5/UFAAAAACIAIPrv3JDIVLeft5JpaP948L/cyobLz9N9fzvM5qNeV34tLGwhAQAAAAAWABQyQn7LBTqGwlw1YR4czWqI99wxNk9PLAAAAAAAFgAUmp9tVa99KP5hNYiVoqw9Q20ISukAAAAAAQErmuf1BQAAAAAiACD679yQyFS3n7eSaWj/ePC/3MqGy8/TfX87zOajXld+LSICAnuk3VLqOOGKWaQyBUAiA0Dkh9fJlvN+voAZ2YE4q6lGRzBEAiBgJOvBFUyD1zFe1cCUi7Q+mabqnDPGwgJMVUzTM+9sqgIgYHaAZKgPoEzzDEc3RZYG1Sxsb4n3eM9OEwpaQaWieTABAQVHUiEDgsVFqH1kh8VRENc3ORR0Nyf7G4DuBuTJxi4ln6dRMTMhAnuk3VLqOOGKWaQyBUAiA0Dkh9fJlvN+voAZ2YE4q6lGUq4AAAA=
>{
  "inputs": [
    {
      "has_utxo": true,
      "is_final": false,
      "missing_sigs": [
        "4ad7e2fc048e440a9dc0b8bc7751df6b04e11cf0"
      ],
      "next": "signer"
    }
  ],
  "fee": 1690,
  "next": "signer"
}

It does seem like our results are slightly different wrt to fee units which is to be expected I guess, but I think we could add the `estimate_vsize`?

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala [hunk 2]
```diff
@@ -154,6 +154,113 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
...
+      PSBTRole.ExtractorPSBTRole
+    } else {
+      (nonWitnessOrUnknownUTXOOpt, witnessUTXOOpt) match {
+        case (None, None) =>
+          PSBTRole.UpdaterPSBTRole
+        case (Some(_), None) | (None, Some(_)) | (Some(_), Some(_)) =>
+          val finalizeT = finalize(txIn)
+          finalizeT match {
+            case Failure(_) =>
+              val dummySigner = Sign.dummySign(ECPublicKey.freshPublicKey)
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:170]: What does this case mean? comments would be helpful, it seems strange to me we are using a dummy sign here.

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala [hunk 3]
```diff
@@ -154,6 +154,113 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
   import org.bitcoins.core.psbt.InputPSBTRecord._
   import org.bitcoins.core.psbt.PSBTInputKeyId._
 
+  def nextRole(txIn: TransactionInput): PSBTRole = {
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:158]: scaladoc

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala [hunk 4]
```diff
@@ -154,6 +154,113 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
...
+                  PSBTRole.FinalizerPSBTRole
+              }
+            case Success(_) =>
+              PSBTRole.FinalizerPSBTRole
+          }
+      }
+    }
+  }
+
+  def prevOutOpt(vout: Int): Option[TransactionOutput] = {
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:188]: scaladoc, just returns thep revious transaction's full output if it's available if I understand correctly?

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala [hunk 5]
```diff
@@ -154,6 +154,113 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
...
+            Some(tx.transactionSpent.outputs(vout))
+          case None => None
+        }
+    }
+  }
+
+  private def missingSigsFromScript(
+      spk: ScriptPubKey): Vector[Sha256Hash160Digest] = {
+    spk match {
+      case EmptyScriptPubKey | _: WitnessCommitment |
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:210]: Hmm.... should we throw in these cases? Is this logic used strictly for analyze or could it be used in other functionality too? 

I guess we probably shouldn't throw in the analyze case, but if we are hitting these script types in the sign case that is a problem.

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala [hunk 6]
```diff
@@ -154,6 +154,113 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
...
+      case None =>
+        nonWitnessOrUnknownUTXOOpt match {
+          case Some(tx) =>
+            Some(tx.transactionSpent.outputs(vout))
+          case None => None
+        }
+    }
+  }
+
+  private def missingSigsFromScript(
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:207]: `@tailrec` and a scaladoc

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala (hunk 1 above)
Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:264]: :+1:

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTRole.scala [hunk 7]
```diff
@@ -0,0 +1,54 @@
+package org.bitcoins.core.psbt
+
+import org.bitcoins.crypto.StringFactory
+
+abstract class PSBTRole {
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/psbt/PSBTRole.scala:5]: Link to the BIP where this is defined please in a scaladoc

**Review by Christewart — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala (hunk 5 above)
Comment (benthecarman) [reply core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:210]: It shouldn't be a problem, if someone had a tx with an input that is an empty spk then it should tell them 0 sigs needed bc it doesn't need sigs. Likely if they do have an empty spk it is us testing something

File: core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala (hunk 6 above)
Comment (benthecarman) [reply core/src/main/scala/org/bitcoins/core/psbt/PSBTMap.scala:207]: Can't do `@tailrec` because of the conditional :(

**Comment by Christewart** [comment]
needs rebase

**Review by Christewart — APPROVED** [review]

**Pull request merged**